### ISSUE TITLE
[tests] tests on Python 3.13 are very slow

### ISSUE THREAD

**Issue by picnixz** [sphinx-doc/sphinx#12174]
The issue is likely due to something happening on 3.13. I did not find a related issue yet but the tests on 3.13 require 15 minutes (!!!)

Possibly related: https://github.com/python/cpython/issues/117108

**Comment by picnixz** [comment]
Some timings (I've simplified the tests but look at this):

```
pytest tests/test_builders/test_build_html_5_output.py::test_html5_output --durations=3
```

### Python 3.10.13

```
0.41s call     tests/test_builders/test_build_html_5_output.py::test_html5_output[images.html-.//img[@src='_images/img.png']-]
0.07s setup    tests/test_builders/test_build_html_5_output.py::test_html5_output[images.html-.//img[@src='_images/img.png']-]
0.01s setup    tests/test_builders/test_build_html_5_output.py::test_html5_output[search.html-.//meta[@name='robots'][@content='noindex']-]
```

### Python 3.13.0a5+ (heads/main:63d6f2623e, Mar 22 2024, 14:15:43) [GCC 7.5.0] on linux (compiled without optimizations)

```
0.92s call     tests/test_builders/test_build_html_5_output.py::test_html5_output[images.html-.//img[@src='_images/img.png']-]
0.14s setup    tests/test_builders/test_build_html_5_output.py::test_html5_output[images.html-.//img[@src='_images/img.png']-]
0.10s setup    tests/test_builders/test_build_html_5_output.py::test_html5_output[search.html-.//meta[@name='robots'][@content='noindex']-]

# with --enable-optimizations (it's slower)
0.98s call     tests/test_builders/test_build_html_5_output.py::test_html5_output[images.html-.//img[@src='_images/img.png']-]
0.14s setup    tests/test_builders/test_build_html_5_output.py::test_html5_output[images.html-.//img[@src='_images/img.png']-]
0.11s setup    tests/test_builders/test_build_html_5_output.py::test_html5_output[search.html-.//meta[@name='robots'][@content='noindex']-]
```

**Comment by picnixz** [comment]
The test slowdown appeared since yesterday so something happened but I'm not sure what. It's apparently not related to the deadsnake version since the latest main branch has this issue.

**Comment by hugovk** [comment]
Yes, this seems to be a CPython thing, our doctests have increased from [~9 mins](https://github.com/python/cpython/actions/runs/8387703237/job/22970424153?pr=117120) to [~40 mins](https://github.com/python/cpython/actions/runs/8386913728/job/22968122197).

**Comment by picnixz** [comment]
Well, good to know (well not good since we wait more). I'll leave it opened until it's patched upstream (to remind myself).

**Comment by picnixz** [comment]
Now that https://github.com/python/cpython/pull/117120 is merged, this should fixed. However, since Sphinx is based on docutils and we could also have a lot of circular references, the memory usage could also blow up (as shown on https://github.com/python/cpython/pull/117120#issuecomment-2015622114).

@hugovk Is there (or is it planned) to be able to disable the incremental gc introduced in Python 3.13 and keep the old gc strategy?

**Comment by chrisjsewell** [comment]
Interesting insight into the garbage collector, and how finely tuned its performance needs to be 😅

**Comment by hugovk** [comment]
> Now that [python/cpython#117120](https://github.com/python/cpython/pull/117120) is merged, this should fixed. However, since Sphinx is based on docutils and we could also have a lot of circular references, the memory usage could also blow up (as shown on [python/cpython#117120 (comment)](https://github.com/python/cpython/pull/117120#issuecomment-2015622114)).

If you see a memory usage, please do report on the issue or create a new one. This is exactly what the alpha releases are for, and thank you for already testing them!

> @hugovk Is it (or is it planned to be) possible to disable the incremental gc introduced in Python 3.13 and keep the old gc strategy?

I don't think so, and don't see it in [What's New](https://docs.python.org/3.13/whatsnew/3.13.html#incremental-garbage-collection) or [changelog](https://docs.python.org/3.13/whatsnew/changelog.html), but I'm not following this area closely.

**Comment by picnixz** [comment]
Quick update: deadsnake version is still the one from yesterday morning so it does not contain the fix yet: `Python 3.13.0a5+ (main, Mar 22 2024, 08:49:13) [GCC 11.4.0]`. Our tests are likely to be back after the next nightly build.

**Comment by picnixz** [comment]
With the new nightly build, tests are back to their usual timings !
